### ISSUE TITLE
question: curl api/create data format?

### ISSUE THREAD

**Issue by Fuzzwah** [dpgaspar/Flask-AppBuilder#310]
I have a super simple model which I'm trying to add a new row via the API:

```
class Card(Model):
    __tablename__ = 'cards'
    id = Column(Integer, primary_key=True)
    number = Column(String(12), nullable=False)
```

I've set up the permissions so Public user has can_add. I can successfully hit the API:

```
$ curl -X GET 'http://localhost:8080/cards/api'
{
  "api_urls": {
    "create": "/cards/api/create",
    "delete": "/cards/api/delete/",
    "read": "/cards/api/read",
    "update": "/cards/api/update/"
  },
  "can_add": true,
  "can_delete": false,
  "can_edit": false,
  "can_show": true,
  "label_columns": {
    "id": "Id",
    "number": "Number"
  },
  "list_columns": [
    "number"
  ],
  "modelview_name": "Cards",
  "modelview_urls": {
    "add": "/cards/add",
    "edit": "/cards/edit/",
    "show": "/cards/show/"
  },
  "order_columns": [
    "number"
  ],
  "page_size": 10,
  "search_fields": {
    "number": "<input class=\"form-control\" id=\"number\" name=\"number\" placeholder=\"Number\" type=\"text\" value=\"\">"
  },
  "search_filters": {
    "number": [
      "Begins with",
      "Ends with",
      "Contains",
      "Equal to",
      "Does not begin with",
      "Does not end with",
      "Does not contain",
      "Not equal to"
    ]
  }
}
```

Here's the results of hitting api/read

```
$ curl -X GET 'http://localhost:8080/cards/api/read'
{
  "count": 1, 
  "label_columns": {
    "id": "Id", 
    "number": "Number"
  }, 
  "list_columns": [
    "number"
  ], 
  "modelview_name": "Cards", 
  "order_columns": [
    "number"
  ], 
  "page": null, 
  "page_size": null, 
  "pks": [
    1
  ], 
  "result": [
    {
      "number": "0123456789"
    }
  ]
}
```

But I'm struggling to get the api/create working:

```
$ curl -X POST -H "Content-type: application/json" -d '{"number": "0123456790"}' http://localhost:8080/cards/api/create
{
  "message": "Invalid form",
  "severity": "warning"
}
```

I'm guessing that my formatting of the json I'm posting is wrong?

I've also tried like this:

```
$ curl -X POST -H "Accept: application/json" -d "number=0123456790" http://localhost:8080/cards/api/create
{
  "message": "Invalid form", 
  "severity": "warning"
}
```

And like this....

```
$ curl -X POST -H "Accept: application/json" --data-binary "number=023456790" http://localhost:8080/cards/api/create
{
    "message": "Invalid form",
    "severity": "warning"
}
```

**Comment by Fuzzwah** [comment]
I dug into views.py to find where the "Invalid form" error message is being generated. It seems that on line 243 it is attempting to do ````form.validate()```` and always fails.

I've hack out this validation in my fork and I'm now able to create new rows with:

  $ curl -X POST -H "Content-type: application/json" -d '{"number": "0123456793"}' http://localhost:8080/cards/api/create
  {
    "message": "Added Row",
    "severity": "success"
  }

**Comment by Fuzzwah** [comment]
I've got to the root cause of all my troubles: wtforms validate() was failing due to csrf.

I've worked out that the option in the config.py of ```CSRF_ENABLED``` is incorrect and should actually be ```WTF_CSRF_ENABLED```.

Now that I have set this to false my post to api/create work as expected.

Here's the key line in the flask_wtf module:

https://github.com/lepture/flask-wtf/blob/master/flask_wtf/csrf.py#L161

Strangely, I can now see that in some of the examples here in flask-appbuilder the testing configs actually do have the correct option name (along with the incorrect one):

https://github.com/dpgaspar/Flask-AppBuilder/search?utf8=%E2%9C%93&q=WTF_CSRF_ENABLED

I'd create a pull request to fix this, but I'm not entirely sure how fabadmin is rigged up to create the default config.py.

Also, I suspect that a better solution would be to csrf_except the API POST end points. I've done some research and attempt to get csrf_except decorator working for api/create, api/delete and api/update but can't get my head around it all.

http://flask-wtf.readthedocs.org/en/latest/csrf.html

**Comment by dpgaspar** [comment]
Yes, i'm not very happy with the REST API for update and and. It's dependent on form validate. We win form validation, but break every compliance regarding REST. I need to rework the API and break some eggs doing it... 

I'll take a look at the decorator also.

**Comment by 7rack** [comment]
@Fuzzwah  i have set `WTF_CSRF_ENABLED = False` . but  i still get 
```json
{
  "message": "Invalid form",
  "severity": "warning"
}
```
do you have any idea ?

**Comment by GregSilverman** [comment]
Hi, I just ran into this problem, too. Not sure if this was an ideal solution, but I changed the two references of `app.config.setdefault('WTF_CSRF_ENABLED', False)` in my virtual env (value was `True`).  Respective files are in `flask_wtf/form.py` and `flask_wtf/csrf.py`.

I would not recommend doing this unless you are either a) only testing, or b) your app is being deployed behind a proxy server (mine is, although I may try one of the above solutions, just for the heck of it).
